rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,315 @@
+| `setModuleMeta` | When compiling ESM to CJS, Babel defined a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |
+| `copyReexports` | When re-exporting an imported value, assume that it's value doesn't change and set it with a simple assignment | `loose` in `transform-modules-commonjs`, `transform-modules-umd`, `transform-modules-amd` _(not documented)_ | | |
+| `newableArrowFunctions` | Do not check that arrow functions are never instantiated using `new` | `spec` in `transform-arrow-functions` | | 5 |
+
+1. `iterableIsArray` is not compatible with `arrayLikeIsIterable`, even if array-like objects could work with `iterableIsArray` in cases where we only rely on indexed access and not on array methods.
+2. With the current implementation, `skipForOfIterationClosing` and `arrayIndexedIteration` should be merged in a single option.
+3. Currently, the `loose` option must be the same for these plugins or it will throw an error. We have a workaround in `@babel/preset-env` to allow setting `loose` there differently from these plugins ([babel/babel#11634](https://github.com/babel/babel/pull/11634))
+4. `?.` also needs to be compiled by the private fields and methods plugins ([babel/babel#11248](https://github.com/babel/babel/pull/11248), but currently there is no way to set it to `loose`
+5. The current default behavior is to consider this assumption as valid, and only produce 100% spec-compliant code when the `spec` option is enabled.
+
+# Drawbacks
+
+- This RFC introduces a very big number of new top-level options (19), and makes it relatively cheap to add new ones. Having a big number of options can add more "tooling fatigue" on the shoulders of our users, and it can make it hard for us to properly document them. However, these new options replace 22 existing plugin options, many of which have the same name (`loose`) but different behaviors.
+
+- ([@JLHwung](https://github.com/JLHwung)) Some `assumptions` are only effective in one plugin, but users will have to go through their presets/plugins to see if they have opt-in to this plugins. Let's say there is an assumption in a stage-1 proposal and the users does not know it is no-op for their config, it may inadvertently turn into a long list of `assumptions` and the no-op assumptions can kick in once users upgrades `preset-env` which may includes it once its stage advances.
```

Comment:
On possible assumptions in early stage plugins, can we limit `assumptions` to be a something that related to stage 3 features or standard spec only? So we don't add assumptions about stage-1 plugins until they are mature enough. The assumptions can still be a valid plugin-specific optionl.

Reply:
I 99%* agree! Maybe we should have the same policy we have for `shippedProposals`?

\* only 99% because, for example, the `#foo in obj` transform is obviously influenced by the assumptions we make when transforming private fields (`privateFieldsAsPublic`).

Reply:
In your case `privateFieldsAsPublic` falls to _shipped_ feature assumptions. A stage-1 proposal implementation can choose to respect such assumptions, but it can not, i.e. register a global `privateInNeverNested` assumption, until it is mature enough -- So other plugins might have to take care of that.

Reply:
100% agree now :smile: 

I'll add a "Policy" section.